array_api_tests/test_statistical_functions.py

```diff
@@ -0,0 +1,365 @@
+        assert math.isnan(out), msg
+    else:
+        msg = f"{out_repr}={out}, should be roughly {expected} [{f_func}]"
+        assert math.isclose(out, expected, rel_tol=0.05), msg
+
+
+@given(
+    x=xps.arrays(
+        dtype=xps.numeric_dtypes(),
+        shape=hh.shapes(min_side=1),
+        elements={"allow_nan": False},
+    ),
+    data=st.data(),
+)
+def test_min(x, data):
+    kw = data.draw(hh.kwargs(axis=axes(x.ndim), keepdims=st.booleans()), label="kw")
+
+    out = xp.min(x, **kw)
+
+    ph.assert_dtype("min", x.dtype, out.dtype)
+    _axes = normalise_axis(kw.get("axis", None), x.ndim)
+    assert_keepdimable_shape(
+        "min", out.shape, x.shape, _axes, kw.get("keepdims", False), **kw
+    )
+    scalar_type = dh.get_scalar_type(out.dtype)
+    for indices, out_idx in zip(axes_ndindex(x.shape, _axes), ah.ndindex(out.shape)):
+        min_ = scalar_type(out[out_idx])
+        elements = []
+        for idx in indices:
+            s = scalar_type(x[idx])
+            elements.append(s)
+        expected = min(elements)
+        assert_equals("min", dh.get_scalar_type(out.dtype), out_idx, min_, expected)
+
+
+@given(
+    x=xps.arrays(
+        dtype=xps.numeric_dtypes(),
+        shape=hh.shapes(min_side=1),
+        elements={"allow_nan": False},
+    ),
+    data=st.data(),
+)
+def test_max(x, data):
+    kw = data.draw(hh.kwargs(axis=axes(x.ndim), keepdims=st.booleans()), label="kw")
+
+    out = xp.max(x, **kw)
+
+    ph.assert_dtype("max", x.dtype, out.dtype)
+    _axes = normalise_axis(kw.get("axis", None), x.ndim)
+    assert_keepdimable_shape(
+        "max", out.shape, x.shape, _axes, kw.get("keepdims", False), **kw
+    )
+    scalar_type = dh.get_scalar_type(out.dtype)
+    for indices, out_idx in zip(axes_ndindex(x.shape, _axes), ah.ndindex(out.shape)):
+        max_ = scalar_type(out[out_idx])
+        elements = []
+        for idx in indices:
+            s = scalar_type(x[idx])
+            elements.append(s)
+        expected = max(elements)
+        assert_equals("max", dh.get_scalar_type(out.dtype), out_idx, max_, expected)
+
+
+@given(
+    x=xps.arrays(
+        dtype=xps.floating_dtypes(),
+        shape=hh.shapes(min_side=1),
+        elements={"allow_nan": False},
+    ),
+    data=st.data(),
+)
+def test_mean(x, data):
+    kw = data.draw(hh.kwargs(axis=axes(x.ndim), keepdims=st.booleans()), label="kw")
+
+    out = xp.mean(x, **kw)
+
+    ph.assert_dtype("mean", x.dtype, out.dtype)
+    _axes = normalise_axis(kw.get("axis", None), x.ndim)
+    assert_keepdimable_shape(
+        "mean", out.shape, x.shape, _axes, kw.get("keepdims", False), **kw
+    )
+    for indices, out_idx in zip(axes_ndindex(x.shape, _axes), ah.ndindex(out.shape)):
+        mean = float(out[out_idx])
+        elements = []
+        for idx in indices:
+            s = float(x[idx])
+            elements.append(s)
+        expected = sum(elements) / len(elements)
```

Comment:
Does this work? I'm a little surprised if the result mean is the same as the naively computed mean. An array library may use an algorithm which is more numerically stable than this (e.g., it could [avoid cancellation](https://en.wikipedia.org/wiki/Kahan_summation_algorithm) in the sum, or by promoting to a higher dtype for intermediate calculations), or even less numerically stable depending on how it is implemented. So I don't think we can test exact values here.

Reply:
Haha yeah, it doesn't quite work yeah. I'm not testing exact values already (my `asset_equals()` uses `isclose()`). I will have to figure out a better/much more lenient method.

Reply:
I've just increased the rel and abs tol, seems fine for NumPy at least. It's pretty lenient now.